storm-client/src/jvm/org/apache/storm/daemon/worker/WorkerState.java

```diff
@@ -488,84 +476,55 @@ public void registerCallbacks() {
         LOG.info("Registering IConnectionCallbacks for {}:{}", assignmentId, port);
         receiver.registerRecv(new DeserializingConnectionCallback(topologyConf,
             getWorkerTopologyContext(),
-            this::transferLocal));
+            this::transferLocalBatch));
     }
 
-    public void transferLocal(List<AddressedTuple> tupleBatch) {
-        Map<Integer, List<AddressedTuple>> grouped = new HashMap<>();
-        for (AddressedTuple tuple : tupleBatch) {
-            Integer executor = taskToShortExecutor.get(tuple.dest);
-            if (null == executor) {
-                LOG.warn("Received invalid messages for unknown tasks. Dropping... ");
-                continue;
-            }
-            List<AddressedTuple> current = grouped.get(executor);
-            if (null == current) {
-                current = new ArrayList<>();
-                grouped.put(executor, current);
-            }
-            current.add(tuple);
-        }
 
-        for (Map.Entry<Integer, List<AddressedTuple>> entry : grouped.entrySet()) {
-            DisruptorQueue queue = shortExecutorReceiveQueueMap.get(entry.getKey());
-            if (null != queue) {
-                queue.publish(entry.getValue());
-            } else {
-                LOG.warn("Received invalid messages for unknown tasks. Dropping... ");
+    private void transferLocalBatch(List<AddressedTuple> tupleBatch) {
+        try {
+            for (int i = 0; i < tupleBatch.size(); i++) {
+                AddressedTuple tuple = tupleBatch.get(i);
+                JCQueue queue = shortExecutorReceiveQueueMap.get(tuple.dest);
```

Comment:
yes  it did show up as a bottleneck during profiling. I recall trying to replace it with an ArrayList... and then backing out as it started getting a bit complicated... i think it had something to do with this map getting periodically refreshed. Let me take a look at it again.